pkg/kube/build_lock.go

```diff
@@ -0,0 +1,415 @@
+		return nil, err
+	} else if !ok {
+		log.Logger().Infof("lock cancelled because not running in tekton")
+		return func() error { return nil }, nil
+	}
+	// Get infos from the headers
+	now := time.Now().UTC().Format(time.RFC3339Nano)
+	owner := os.Getenv("REPO_OWNER")
+	if owner == "" {
+		log.Logger().Warnf("no REPO_OWNER provided")
+		return nil, fmt.Errorf("no REPO_OWNER provided")
+	}
+	repository := os.Getenv("REPO_NAME")
+	if repository == "" {
+		log.Logger().Warnf("no REPO_NAME provided")
+		return nil, fmt.Errorf("no REPO_NAME provided")
+	}
+	branch := os.Getenv("BRANCH_NAME")
+	if branch == "" {
+		log.Logger().Warnf("no BRANCH_NAME provided")
+		return nil, fmt.Errorf("no BRANCH_NAME provided")
+	}
+	build := os.Getenv("BUILD_NUMBER")
+	if _, err := strconv.Atoi(build); err != nil {
+		log.Logger().Warnf("no BUILD_NUMBER provided: %s\n", err.Error())
+		return nil, err
+	}
+	// Find our pod
+	podList, err := kubeClient.CoreV1().Pods(devNamespace).List(metav1.ListOptions{
+		LabelSelector: fmt.Sprintf("owner=%s,repository=%s,branch=%s,build=%s,jenkins.io/pipelineType=build", owner, repository, branch, build),
+	})
+	if err != nil {
+		return nil, err
+	} else if len(podList.Items) != 1 {
+		return nil, fmt.Errorf("%d pods found for this job (owner=%s,repository=%s,branch=%s,build=%s,jenkins.io/pipelineType=build)",
+			len(podList.Items), owner, repository, branch, build)
+	}
+	pod := &podList.Items[0]
+	// kubernetes library seems to forget APIVersoin and Kind
+	// fill those if they're missing
+	if pod.APIVersion == "" {
+		pod.APIVersion = "v1"
+	}
+	if pod.Kind == "" {
+		pod.Kind = "Pod"
+	}
+	podKind := pod.Kind
+	// Create the lock object
+	lock := &v1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      fmt.Sprintf("jx-lock-%s", namespace),
+			Namespace: devNamespace,
+			Labels: map[string]string{
+				"namespace":  namespace,
+				"owner":      owner,
+				"repository": repository,
+				"branch":     branch,
+				"build":      build,
+			},
+			Annotations: map[string]string{
+				"jenkins-x.io/created-by": "Jenkins X",
+				"warning":                 "DO NOT REMOVE",
+				"purpose": fmt.Sprintf("This is a deployment lock for the "+
+					"namespace \"%s\". It prevents several deployments to "+
+					"edit the same namespace at the same time. It will "+
+					"automatically be removed once the deployemnt is "+
+					"finished, or replaced by the next deployemnt to run.",
+					namespace),
+			},
+			OwnerReferences: []metav1.OwnerReference{{
+				APIVersion: pod.APIVersion,
+				Kind:       pod.Kind,
+				Name:       pod.Name,
+				UID:        pod.UID,
+			}},
+		},
+		Data: map[string]string{
+			"namespace":  namespace,
+			"owner":      owner,
+			"repository": repository,
+			"branch":     branch,
+			"build":      build,
+			"pod":        pod.Name,
+			"timestamp":  now,
+		},
+	}
+	for k, v := range buildLockLabels {
+		lock.Labels[k] = v
+	}
+	// this loop continuously tries to create the lock
+Create:
```

Comment:
hmmm, do we really need to use labels?

Reply:
We never have to use labels (same as you never have to use GOTO). But in that case, we will instead have to signify which loop should break and which one should continue. I don't thin kit will make the algorithm more clear. The algorithm is basically a state machine, and those are well described with GOTO ...

Reply:
Oh sorry did you mean labels on the lock ? Well most of them let you know who is locking currently. They are similar to the ones in the pod. I also added a label to be sure that build are not log by another configmap that shouldn't be here (or maybe later another version of the same lock), that would cause a bug due to a different format. Nothing mandatory, just felt like it was cleaner.